### PR TITLE
Enhance Tests of Source and New Tests of Sink Plugin

### PR DESCRIPTION
**Pull request by Vipinofficial11**

### CODE REVIEW

**Comment by Vipinofficial11** [comment]
This PR has all the Updates For Source and Multi Source e2e Tests based on the Framework changes and also there is e2e tests for Service Now Sink. (FYI : Step to validate data written to sink is also added.)

@rmstar @itsankit-google please review.

**Review by rmstar — COMMENTED** [review]
Please take a look at test failures

File: src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIClient.java [hunk 1]
```diff
@@ -18,6 +18,7 @@
 
 import org.apache.http.client.methods.CloseableHttpResponse;
 import org.apache.http.client.methods.HttpGet;
+import org.apache.http.client.methods.HttpPatch;
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIClient.java]: not needed?

File: src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java [hunk 2]
```diff
@@ -39,6 +39,7 @@ public class RestAPIResponse {
   private static List<Integer> successCodes = new ArrayList<Integer>() {
     {
       add(HttpStatus.SC_OK);
+      add(HttpStatus.SC_CREATED);
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java]: unused?

File: src/main/java/io/cdap/plugin/servicenow/apiclient/ServiceNowTableAPIClientImpl.java [hunk 3]
```diff
@@ -147,19 +145,21 @@ public List<Map<String, Object>> fetchTableRecords(String tableName, SourceValue
    * @param tableName ServiceNow Table name
    * @param entity    Details of the Record to be created
    */
-  public void createRecord(String tableName, HttpEntity entity) throws IOException {
+  public String createRecord(String tableName, HttpEntity entity) throws IOException {
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/servicenow/apiclient/ServiceNowTableAPIClientImpl.java:150]: Annotate with @VisibleForTesting

File: src/main/java/io/cdap/plugin/servicenow/apiclient/ServiceNowTableAPIClientImpl.java [hunk 4]
```diff
@@ -169,8 +169,48 @@ public void createRecord(String tableName, HttpEntity entity) throws IOException
...
+    return result.get(ServiceNowConstants.SYSTEM_ID).getAsString();
   }
 
+  /**
+   * Return a record from ServiceNow application.
+   *
+   * @param tableName The ServiceNow table name
+   * @param query     The query
+   */
+  public Map<String, String> getRecordFromServiceNowTable(String tableName, String query)
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/servicenow/apiclient/ServiceNowTableAPIClientImpl.java:197]: @VisibleForTesting

File: src/e2e-test/java/io/cdap/plugin/servicenowsink/actions/ServiceNowSinkPropertiesPageActions.java [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+        return false;
+      }
+    }
+    return result;
+  }
+
+  /**
+   * Return BigQuery supported date formats.
+   */
+  private static Date checkBigQueryDateFormat(String value) {
```
Comment (rmstar) [inline src/e2e-test/java/io/cdap/plugin/servicenowsink/actions/ServiceNowSinkPropertiesPageActions.java:140]: @Nullable

File: src/e2e-test/java/io/cdap/plugin/servicenow/stepsdesign/DesignTimeSteps.java [hunk 6]
```diff
@@ -16,10 +16,13 @@
 
 package io.cdap.plugin.servicenow.stepsdesign;
 
+import io.cdap.e2e.pages.locators.CdfSchemaLocators;
+import io.cdap.e2e.utils.AssertionHelper;
```
Comment (rmstar) [inline src/e2e-test/java/io/cdap/plugin/servicenow/stepsdesign/DesignTimeSteps.java]: where are these imports used?

File: src/e2e-test/java/io/cdap/plugin/bigquery/stepsdesign/BigQueryCommonSteps.java [hunk 7]
```diff
@@ -19,13 +19,15 @@
...
 import io.cdap.e2e.utils.BigQueryClient;
+import io.cdap.e2e.utils.PluginPropertyUtils;
 import io.cdap.plugin.tests.hooks.TestSetupHooks;
 import io.cucumber.java.en.Then;
 import io.cucumber.java.en.When;
 import org.apache.commons.lang3.RandomStringUtils;
 import org.junit.Assert;
 
 import java.io.IOException;
+import java.util.Optional;
```
Comment (rmstar) [inline src/e2e-test/java/io/cdap/plugin/bigquery/stepsdesign/BigQueryCommonSteps.java]: unused?

File: src/e2e-test/features/servicenowsource/RunTimeWithMacros.feature [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+    And Enter runtime argument value "receiving_slip_line" for key "tableName"
+    And Enter runtime argument value from environment variable "client.id" for key "clientId"
+    And Enter runtime argument value from environment variable "client.secret" for key "clientSecret"
+    And Enter runtime argument value from environment variable "rest.api.endpoint" for key "restApiEndpoint"
+    And Enter runtime argument value from environment variable "pipeline.user.username" for key "username"
+    And Enter runtime argument value from environment variable "pipeline.user.password" for key "password"
+    And Run the Pipeline in Runtime with runtime arguments
+    And Wait till pipeline is in running state
+    And Open and capture logs
+    Then Verify the pipeline status is "Succeeded"
```
Comment (rmstar) [inline src/e2e-test/features/servicenowsource/RunTimeWithMacros.feature:88]: check that data is written correctly to sink

File: src/e2e-test/features/servicenowsource/RunTime.feature [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    And Enter input plugin property: "datasetProject" with value: "datasetprojectId"
+    And Configure BigQuery sink plugin for Dataset and Table
+    Then Validate "BigQuery" plugin properties
+    And Close the Plugin Properties page
+    And Save and Deploy Pipeline
+    And Run the Pipeline in Runtime
+    And Wait till pipeline is in running state
+    And Open and capture logs
+    And Verify the pipeline status is "Succeeded"
+    And Verify count of no of records transferred to the target BigQuery Table
```
Comment (rmstar) [inline src/e2e-test/features/servicenowsource/RunTime.feature:70]: in addition to count, verify that correct data is written to sink

File: src/e2e-test/features/servicenowsink/RunTimeWithMacros.feature [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+    And Enter runtime argument value "receiving_slip_line" for key "tableName"
+    And Enter runtime argument value from environment variable "client.id" for key "clientId"
+    And Enter runtime argument value from environment variable "client.secret" for key "clientSecret"
+    And Enter runtime argument value from environment variable "rest.api.endpoint" for key "restApiEndpoint"
+    And Enter runtime argument value from environment variable "pipeline.user.username" for key "username"
+    And Enter runtime argument value from environment variable "pipeline.user.password" for key "password"
+    And Run the Pipeline in Runtime with runtime arguments
+    And Wait till pipeline is in running state
+    And Verify the pipeline status is "Succeeded"
+    Then Open and capture logs
```
Comment (rmstar) [inline src/e2e-test/features/servicenowsink/RunTimeWithMacros.feature:162]: verify that correct data is written to sink

File: src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIClient.java (hunk 1 above)
Comment (Vipinofficial11) [reply src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIClient.java]: yes it was not needed, removed it.

File: src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java (hunk 2 above)
Comment (Vipinofficial11) [reply src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java]: yes, removed.

File: src/e2e-test/java/io/cdap/plugin/servicenowsink/actions/ServiceNowSinkPropertiesPageActions.java (hunk 5 above)
Comment (Vipinofficial11) [reply src/e2e-test/java/io/cdap/plugin/servicenowsink/actions/ServiceNowSinkPropertiesPageActions.java:140]: Added.

File: src/e2e-test/java/io/cdap/plugin/servicenow/stepsdesign/DesignTimeSteps.java (hunk 6 above)
Comment (Vipinofficial11) [reply src/e2e-test/java/io/cdap/plugin/servicenow/stepsdesign/DesignTimeSteps.java]: They were unused, removed.

File: src/main/java/io/cdap/plugin/servicenow/apiclient/ServiceNowTableAPIClientImpl.java (hunk 3 above)
Comment (Vipinofficial11) [reply src/main/java/io/cdap/plugin/servicenow/apiclient/ServiceNowTableAPIClientImpl.java:150]: I am actually using this method in the **e2e-test** package.(If I will use the Annotation Ill be only able to use it in Test classes [Junit tests classes]).

File: src/main/java/io/cdap/plugin/servicenow/apiclient/ServiceNowTableAPIClientImpl.java (hunk 4 above)
Comment (Vipinofficial11) [reply src/main/java/io/cdap/plugin/servicenow/apiclient/ServiceNowTableAPIClientImpl.java:197]: I am actually using this method in the e2e-test package.

File: src/e2e-test/java/io/cdap/plugin/bigquery/stepsdesign/BigQueryCommonSteps.java (hunk 7 above)
Comment (Vipinofficial11) [reply src/e2e-test/java/io/cdap/plugin/bigquery/stepsdesign/BigQueryCommonSteps.java]: yes, removed

File: src/e2e-test/features/servicenowsink/RunTimeWithMacros.feature (hunk 10 above)
Comment (Vipinofficial11) [reply src/e2e-test/features/servicenowsink/RunTimeWithMacros.feature:162]: Added the step. (d4426a7b658927fc2fdefd4b5e578521bd150bcf)

File: src/e2e-test/features/servicenowsource/RunTimeWithMacros.feature (hunk 8 above)
Comment (Vipinofficial11) [reply src/e2e-test/features/servicenowsource/RunTimeWithMacros.feature:88]: This PR is actually mainly for the ServiceNow Sink e2e tests, I have just enhanced the Source and Multi Source steps based on the latest changes in the framework. I will be raising a separate for Source and Multi Source data validation after this one.

File: src/e2e-test/features/servicenowsource/RunTime.feature (hunk 9 above)
Comment (Vipinofficial11) [reply src/e2e-test/features/servicenowsource/RunTime.feature:70]: Same reason as above, I will raise a Separate PR for this.

File: src/main/java/io/cdap/plugin/servicenow/apiclient/ServiceNowTableAPIClientImpl.java [hunk 11]
```diff
@@ -204,8 +206,8 @@ public Map<String, String> getRecordFromServiceNowTable(String tableName, String
     requestBuilder.setAuthHeader(accessToken);
     apiResponse = executeGet(requestBuilder.build());
     JsonObject jsonObject = gson.fromJson(apiResponse.getResponseBody(), JsonObject.class);
-    JsonArray jsonResultArray = jsonObject.getAsJsonArray(ServiceNowConstants.RESULT);
-    Map<String, String> responseMap = gson.fromJson(jsonResultArray.get(0), Map.class);
+    serviceNowJsonResultArray = jsonObject.getAsJsonArray(ServiceNowConstants.RESULT);
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/servicenow/apiclient/ServiceNowTableAPIClientImpl.java:209]: why did we need this change?

File: src/e2e-test/java/io/cdap/plugin/servicenow/stepsdesign/DesignTimeSteps.java [hunk 12]
```diff
@@ -16,6 +16,7 @@
 
 package io.cdap.plugin.servicenow.stepsdesign;
 
+
```
Comment (rmstar) [inline src/e2e-test/java/io/cdap/plugin/servicenow/stepsdesign/DesignTimeSteps.java]: remove extra newline

File: src/main/java/io/cdap/plugin/servicenow/apiclient/ServiceNowTableAPIClientImpl.java (hunk 11 above)
Comment (Vipinofficial11) [reply src/main/java/io/cdap/plugin/servicenow/apiclient/ServiceNowTableAPIClientImpl.java:209]: ServiceNow response is being stored in this variable, so using this in the validation logic.

File: src/e2e-test/java/io/cdap/plugin/servicenowsink/actions/ServiceNowSinkPropertiesPageActions.java [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+import java.util.Map;
+import java.util.Set;
+import javax.annotation.Nullable;
+
+/**
+ * ServiceNow sink - Properties page - Actions.
+ */
+
+public class ServiceNowSinkPropertiesPageActions {
+  private static final int DEFAULT_SCALE = 2;
```
Comment (rmstar) [inline src/e2e-test/java/io/cdap/plugin/servicenowsink/actions/ServiceNowSinkPropertiesPageActions.java]: use this - https://github.com/data-integrations/servicenow-plugins/blob/develop/src/main/java/io/cdap/plugin/servicenow/util/ServiceNowConstants.java#L268

File: src/e2e-test/java/io/cdap/plugin/servicenowsink/actions/ServiceNowSinkPropertiesPageActions.java [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+  public static ServiceNowSourceConfig config;
+  public static Map<String, String> responseFromServiceNowTable;
```
Comment (rmstar) [inline src/e2e-test/java/io/cdap/plugin/servicenowsink/actions/ServiceNowSinkPropertiesPageActions.java]: why do these need to be public?

File: src/e2e-test/java/io/cdap/plugin/servicenow/stepsdesign/DesignTimeSteps.java (hunk 12 above)
Comment (Vipinofficial11) [reply src/e2e-test/java/io/cdap/plugin/servicenow/stepsdesign/DesignTimeSteps.java]: removed.(04c5fcb85a4ffccdcfd37580d88f6ae267c0f49e)

File: src/e2e-test/java/io/cdap/plugin/servicenowsink/actions/ServiceNowSinkPropertiesPageActions.java (hunk 13 above)
Comment (Vipinofficial11) [reply src/e2e-test/java/io/cdap/plugin/servicenowsink/actions/ServiceNowSinkPropertiesPageActions.java]: Updated.(04c5fcb85a4ffccdcfd37580d88f6ae267c0f49e)

File: src/e2e-test/java/io/cdap/plugin/servicenowsink/actions/ServiceNowSinkPropertiesPageActions.java (hunk 14 above)
Comment (Vipinofficial11) [reply src/e2e-test/java/io/cdap/plugin/servicenowsink/actions/ServiceNowSinkPropertiesPageActions.java]: Not needed as public, Changed to private.(https://github.com/data-integrations/servicenow-plugins/pull/35/commits/04c5fcb85a4ffccdcfd37580d88f6ae267c0f49e#diff-cc3534f5337d36a161fd6bddf24a37827ae2f6ff149c7f26a36a0346f846b118R118)

**Review by rmstar — APPROVED** [review]
please squash commits

**Comment by Vipinofficial11** [comment]
> please squash commits

commits squashed.

**Pull request merged**